### PR TITLE
Worse is better

### PR DESCRIPTION
**Pull request by Drup**
When we created functoria, I had this fancy idea of using graphs everywhere.
With hindsight, It wasn't such a good idea.

This commit remove all the formal graph structure, and use DAGs implemented as terms. The type of `Impl.t` is expanded to cover all we need, and most of the pre-computations are done on `Impl.t` directly. `Device_graph.t` now only contain devices, and a new function `Impl.eval` removes everything except the devices, which are used for code output.

All this removes quite a bit of code, and more importantly, quite a lot of `assert false` which have been the cause of numerous issues ...

I only have tested this very loosely (using the testsuite), so I would appreciate if someone can run this on their favorite unikernel. The code output *will* be different (because the order and the ids are not particularly stable), but it should be equivalent.

Unfortunately, I had to touch `Lib` in the process, in order to insert the calls to `Impl.eval` appropriately. I seem to have changed something on the way (see the test change in the last commit). I'm not sure what.

The dot output is not implemented yet, I'll do that next.

### CODE REVIEW

File: lib/functoria/impl.ml
Comment (Drup) [inline lib/functoria/impl.ml:203]: This is rather dirty. If anyone has a proposal to make that safe in a concise manner. I'm all ears. So far my plan is to use equality witnesses.

**Comment by dinosaure** [comment]
I'm not really aware about this codebase but:
1) Travis CI try to compile `mirage-skeleton`, as far as Travis CI is green, we can say that this PR is fine for most of our cases
2) I can play a bit with some of my unikernels
3) it can be interesting to integrate [hannesm/functoria-playground](https://github.com/hannesm/functoria-playground) (/cc @hannesm) and see if such cases work (I suspect that these _unikernels_ reached the same `assert false` as me)
4) I'm wondering about MirageOS 4 and MirageOS 3.10. Currently, the merge will be apply on MirageOS 4 and we probably backport this patch on MirageOS 3 (branch 3 but I'm not sure)

Otherwise, `hmap` is missing.

File: lib/functoria/impl.ml
Comment (dinosaure) [reply lib/functoria/impl.ml:203]: Please use equality witnesses 👍

Comment (Drup) [reply lib/functoria/impl.ml:206]: Done!

**Comment by Drup** [comment]
dot output is implemented, and I added equality witnesses to implement the polymorphic hash consing on `impl`s in a safe way. It seems to work well.

We still have 3 walks over `impl`s. The typing gets in the way of simplifying/factorizing further. I though about getting rid of preserving sharing on `impl`s, and only do that on device graphs (where it's essential, to avoid double inits), but that would result in horrible dots and compromise collections. Maybe someone else with fresh eyes will have an idea.

The last commit changes a test: the last version of dune doesn't create `.merlin` files by default anymore.

I think it's ready for review.

**Comment by dinosaure** [comment]
Thanks, I will do the end-to-end test by hands on some unikernels with UNIX and `hvt` and let Travis CI tell to us that all works :+1:. I will figure out how to patch-back your PR into MirageOS 3.10 (and cut a release with that)

**Comment by samoht** [comment]
I like the approach very much. Give me a few days to review the code in details before merging :-)

File: lib/functoria/device.mli
Comment (samoht) [inline lib/functoria/device.mli:54]: Maybe move `Eq` to a top-level module, that's not very nice to have `Misc` leaking in the public API.

File: lib/functoria/impl.ml
Comment (samoht) [inline lib/functoria/impl.ml:201]: Can you add a bit of documentation here? Especially the logics to decide whether to use default condition values depending  on `partial` and which key is in `context`. It always take me a while to remember why it's working like this :-)

File: lib/functoria/impl.ml
Comment (samoht) [inline lib/functoria/impl.ml:238]: That `eval` function is so simple now :-) 💯 ❤️

File: lib/functoria/impl.mli
Comment (samoht) [inline lib/functoria/impl.mli:94]: did you remove `map` on purpose?

File: lib/functoria/lib.ml
Comment (samoht) [inline lib/functoria/lib.ml:64]: you probably don't need to rename that function

File: lib/functoria/lib.ml
Comment (samoht) [inline lib/functoria/lib.ml:114]: same here, `eval_cached` still works fine on that context I think

File: lib/functoria/lib.ml
Comment (samoht) [inline lib/functoria/lib.ml:136]: Is it possible to continue to show something sensible in text mode (instead of `Fmt.nop`) ?

File: lib/functoria/misc.ml
Comment (samoht) [inline lib/functoria/misc.ml:104]: `hash` is a bit of a weird name here -- as we ensure that we return a unique ID. Maybe just call it `id`? 

Or maybe don't use it for equality (as it is done in one other part of the code -- I think when trying to equalise devices).

File: lib/functoria/misc.mli
Comment (samoht) [inline lib/functoria/misc.mli:42]: do we need to expose `andeq` and `andb` ?

File: test/functoria/test_graph.ml
Comment (samoht) [inline test/functoria/test_graph.ml:35]: I am confused: why is this different now? How do we ensure devices have unique names (and variable names)?

**Comment by samoht** [comment]
Ok so: many thanks for doing this, it's a big cleanup on the existing code and simplifies a lot of things. I'm looking forward to merge it asap :-)

The only thing that I'm not super sure is the changes in `var_name`. I guess the previous generated IDs were the for all the nodes in the impl graphs while the new ones are only for the nodes in the device graph? But why do the test are re-using the same IDs in different variables?

File: lib/functoria/impl.mli
Comment (Drup) [reply lib/functoria/impl.mli:94]: Yes, it's not used outside the module now.

File: lib/functoria/lib.ml
Comment (Drup) [reply lib/functoria/lib.ml:64]: Fair enough. I though the name was a bit misleading, since it doesn't call `eval`.

File: lib/functoria/lib.ml
Comment (Drup) [reply lib/functoria/lib.ml:136]: `Config.pp` was defined as `Device_graph.pp` ... which was `Fmt.nop`. It was printing nothing all along.

Comment (Drup) [reply lib/functoria/lib.ml:136]: We could technically use `Impl.pp` here, but that's more of a debug output than something you want to show to users.

File: lib/functoria/misc.mli
Comment (Drup) [reply lib/functoria/misc.mli:42]: They are used in the implementation of `Impl.equal`

File: lib/functoria/impl.ml
Comment (Drup) [reply lib/functoria/impl.ml:201]: I flipped the bool to `simplify ~full` at its slighly more intuitive, and added a comment.

File: lib/functoria/device.mli
Comment (Drup) [reply lib/functoria/device.mli:54]: Done, it's called `Typeid`

File: lib/functoria/misc.ml
Comment (Drup) [reply lib/functoria/misc.ml:104]: Done, renamed to `id`.

File: lib/functoria/misc.mli
Comment (Drup) [reply lib/functoria/misc.mli:42]: aaand, they are gone.

File: test/functoria/test_graph.ml
Comment (Drup) [reply test/functoria/test_graph.ml]: Okay, so the point of the test was indeed rendered a bit moot. I modified it partially: Now, we create a full implementation, pass it through the eval function, ask for a topo sort and tests the IDs on that. What do you think of that ?

**Comment by Drup** [comment]
I think I solved all your points, I also added documentations in various places.

**Review by samoht — APPROVED** [review]

**Comment by samoht** [comment]
All green! @dinosaure did you test it on mirage-skeleton already? Happy to merge that PR once that works ok.

**Comment by TheLortex** [comment]
I've set up a temporary mirage CI pipeline to test the PR: https://ci.mirage.io/main/pipeline.html
The outcome is that mirage-skeleton fails on the same places as with the main branch (except for issues that should be fixed if you rebase your PR on master to pick up the latest fixes).

**Comment by Drup** [comment]
@TheLortex Thanks! From what I can see, those failure are not related to the current PR. Is that correct ? Otherwise, I just rebased, so we can have a clearer picture.

**Comment by TheLortex** [comment]
Yes, now we can see that the errors are the same on `master` and with your PR, so this is looking good !

**Comment by dinosaure** [comment]
Ready to merge, we will discuss internally if we want to patch-back that in MirageOS 3 or not, thanks your great work both 👍 !

**Comment by Drup** [comment]
Thanks, I'm happy it turned out so well, this code has been bothering me for a few years now, and now hopefully it will be more reliable, and people will be able to look and modify it!

File: lib/functoria/engine.ml
Comment (samoht) [inline lib/functoria/engine.ml:94]: @Drup are you sure we really want to select all devices with the same id here? I don't remember why we need to need to find "all devices" here. The `connect` code mainly try to associate the init devices by their counterpart in the graph.

Comment (Drup) [reply lib/functoria/engine.ml:94]: `find_all_devices` was added rather recently with the `~init` parameter. I'm personally not super convinced we need to code it like that, but I wasn't sure, so I preserved the same semantics as was present before.

Since we are very careful never to duplicate any devices, I think this list always ever have one element ...

Comment (samoht) [reply lib/functoria/engine.ml:94]: but what happens if you add the same device twice in the graph?

Comment (samoht) [reply lib/functoria/engine.ml:94]: That function should probably just be a call to `with_left_most_device` ...

**Pull request merged**
